internal/provider/resource_credential.go

```diff
@@ -3,107 +3,146 @@ package provider
 				Description: "Add credentials to the client",
-				Type:        schema.TypeBool,
 				Optional:    true,
-				Default:     false,
+				Computed:    true,
+				Default:     booldefault.StaticBool(false),
 			},
-			"controller_credential": {
+			"controller_credential": schema.BoolAttribute{
 				Description: "Add credentials to the controller",
-				Type:        schema.TypeBool,
 				Optional:    true,
-				Default:     true,
+				Computed:    true,
+				Default:     booldefault.StaticBool(true),
 			},
-			"name": {
+			"name": schema.StringAttribute{
 				Description: "The name to be assigned to the credential",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
 		},
 	}
 }
 
-func AttributeEntryToString(input interface{}) string {
-	switch t := input.(type) {
-	case bool:
-		return strconv.FormatBool(t)
-	case int64:
-		return strconv.FormatInt(t, 10)
-	case float64:
-		return strconv.FormatFloat(t, 'f', 0, 64)
-	default:
-		return input.(string)
-	}
-}
-
-func resourceCredentialCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+func (c *credentialResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan credentialResourceModel
 
-	var diags diag.Diagnostics
+	// Read Terraform configuration from the request into the resource model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
 
-	attributesRaw := d.Get("attributes").(map[string]interface{})
-	authType := d.Get("auth_type").(string)
-	clientCredential := d.Get("client_credential").(bool)
-	cloud := d.Get("cloud").([]interface{})
-	controllerCredential := d.Get("controller_credential").(bool)
-	credentialName := d.Get("name").(string)
+	// Get the resource fields
+	var attributesRaw map[string]interface{}
+	var authType string
+	var clientCredential bool
+	var cloud []interface{}
+	var controllerCredential bool
+	var credentialName string
+
+	plan.Attributes.ElementsAs(ctx, attributesRaw, false)
+	authType = plan.AuthType.ValueString()
+	clientCredential = plan.ClientCredential.ValueBool()
+	plan.Cloud.ElementsAs(ctx, cloud, false)
+	controllerCredential = plan.ControllerCredential.ValueBool()
+	credentialName = plan.Name.ValueString()
 
 	attributes := make(map[string]string)
 	for key, value := range attributesRaw {
 		attributes[key] = AttributeEntryToString(value)
 	}
-	response, err := client.Credentials.CreateCredential(juju.CreateCredentialInput{
+	// Prevent a segfault if client is not yet configured
+	if c.client == nil {
+		resp.Diagnostics.AddError(
+			"Provider Error, Client Not Configured",
+			"Unable to create credential resource. Expected configured Juju Client. "+
+				"Please report this issue to the provider developers.",
+		)
+		return
+	}
```

Comment:
Do this first in the functions, no use getting the plan etc if the client is nil.